design/aws-ecs-priv-registry-support.md

```diff
@@ -0,0 +1,146 @@
+# AWS ECS - Support for Private Registry Authentication
+
+To address issue [#1698](https://github.com/awslabs/aws-cdk/issues/1698), the ECS construct library should provide a way for customers to specify [`repositoryCredentials`](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerDefinition.html#ECS-Type-ContainerDefinition-repositoryCredentials) on their container. 
+
+Minimally, this would mean adding a new string field on `ContainerDefinition`, however this doesn't provide any added value in terms of logical grouping or resource creation. We can instead modify the existing ECS CDK construct [`ContainerImage`](https://github.com/awslabs/aws-cdk/blob/master/packages/%40aws-cdk/aws-ecs/lib/container-image.ts) so that repository credentials are specified along with the image they're meant to access.
+
+## General approach
+
+The [`ecs.ContainerImage`](https://github.com/awslabs/aws-cdk/blob/master/packages/%40aws-cdk/aws-ecs/lib/container-image.ts) class already includes constructs for 3 types of images:
+
+* DockerHubImage
```

Comment:
I have been told that `DockerHubImage` is a misnomer, since the same class can be used for *any* image that is identifiable to Docker by simply specifying a name. And in fact, you could already put a registry into this and it wouldn't complain:

```ts
ContainerImage.fromDockerHub('my.registry.com/image')  // Waitaminit...
```

Thinking it through purely from the consumer's point of view (disregarding implementation for a moment), maybe we just need to generalize this class? We could take properties, and make something like this:

```ts
ContainerImage.fromInternet('my.registry.com/image', {
  username: 'username',
  password: 'password'
});

// Or even
ContainerImage.fromInternet('image', {
  registry: 'my.registry.com',
  username: 'username',
  password: 'password'
});
```

Maybe we even model the registry more explicitly. Feels like it might make sense to have an object that represents the "private registry" directly?

```ts
const registry = new ecs.ContainerRegistry({
  registry: 'my.registry.com',
  username: 'username',
  password: 'password'
});

td.addContainer(..., {
  image: registry.getImage('image')
});
```

I suppose the latter would play better with ECR's version of private registries too?

What I'm saying is: none of the current API is holy, and we're still at liberty to shift it around as much as we like to make it make the most sense.

Reply:
I also think `fromInternet` is not accurate because you could also specify images that were build locally. Maybe it could be `ContainerImage.fromName`? or `withName` or `byName`..

Reply:
Well but you're never *retrieving* them from your local Docker, right?

Or do you mean, a `ContainerImage` that--by some means--already exists on your machine and gets pushed to ECR? So same as an `AssetImage` but without the `docker build`? Because I don't think we have support for that at all. At least `fromDockerHub()` is not going to do that for you. Would be a good feature though...

Reply:
You're right re: `DockerHubImage` being a misnomer; it would work with any image URI so long as creds were provided for private repos. 

Generalizing it to accommodate the `PrivateImage` class would work and reduce the # of subclasses, I just wasn't sure renaming was an option (also: `fromDockerHub` has the advantage of helping people identify a common use case). Maybe we can callout "DockerHub" in the construct class/docs to keep this association explicit?